File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/HomeFeedRules.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.GenreItem
import com.bockmedia.console.data.api.dto.PlaylistSummary
import com.bockmedia.console.data.api.dto.StreamHistoryItem
import kotlin.random.Random

/**
 * Client-side rules for Spotify-style home rows (until a dedicated /api/home exists).
 *
 * Jump back in — recent listening (playlists, albums, artists) plus recently added playlists.
 * Top mixes — genre / era style mixes from your library and listening stats (familiar-heavy).
 * Daily mixes — genre clusters (smart or named daily playlists), favorites blended with similar tracks.
 * Recent playlists — any playlist you've played, ordered by recency (not “created by you” only).
 * Radio — infinite shuffle seeded from one artist, song, or genre (not finite mix playlists).
 * Discover — new-to-you leaning playlists and artists you haven't played recently.
 */
object HomeFeedRules {
    private val dailyMixPattern = Regex("(?i)daily mix|daylist")
    private val discoverPattern = Regex("(?i)discover weekly|new release|fresh find|new to you")
    /** Word-boundary mix markers — avoids false positives like "Generation" (era) or "Remix" substring-only. */
    private val genreMixPattern = Regex(
        "(?i)\\bmix\\b|\\bmixes\\b|\\bremix\\b|\\bremixes\\b|essentials|" +
            "\\bdecade\\b|\\bera\\b|\\bhits\\b|\\bparty\\b|\\bfocus\\b|\\bfavorites\\b",
    )
    private val mixLikeNamePattern = Regex("(?i)\\bmix\\b|\\bmixes\\b|\\bremix\\b|\\bremixes\\b")
    private val explicitRadioPlaylistPattern = Regex("(?i)\\bradio\\b|\\bstation\\b")
    private val mixLikePlaylistPattern = Regex("(?i)\\bmix\\b|daily|discover weekly|essentials|station")

    fun isDailyMixName(name: String): Boolean = dailyMixPattern.containsMatchIn(name)

    fun isDiscoverName(name: String): Boolean = discoverPattern.containsMatchIn(name)

    fun isGenreMixPlaylistName(name: String, genre: String? = null): Boolean {
        if (isDailyMixName(name) || isDiscoverName(name)) return false
        if (!genreMixPattern.containsMatchIn(name)) return false
        return genre == null || nameContainsGenre(name, genre)
    }

    /** True when the playlist name looks mix/remix-like (not bare "Remix" substring inside unrelated words). */
    fun hasMixLikeName(name: String): Boolean = mixLikeNamePattern.containsMatchIn(name)

    fun nameContainsGenre(name: String, genre: String): Boolean {
        val g = genre.trim()
        if (g.isBlank()) return false
        if (name.contains(g, ignoreCase = true)) return true
        // Multi-word genres: require every token to appear (e.g. "Hip Hop" in playlist title).
        val tokens = g.split(Regex("\\s+")).filter { it.length > 1 }
        return tokens.size > 1 && tokens.all { name.contains(it, ignoreCase = true) }
    }

    /** Best library playlist for a top-genre mix row — prefers exact "${genre} Mix" over loose matches. */
    fun bestGenreMixPlaylist(all: List<PlaylistSummary>, genre: String): PlaylistSummary? {
        val g = genre.trim()
        if (g.isBlank()) return null
        return all.filter { it.tracks > 0 && isGenreMixPlaylistName(it.name, g) }
            .maxByOrNull { genreMixNameScore(it.name, g) * 10_000 + it.tracks }
    }

    private fun genreMixNameScore(name: String, genre: String): Int {
        var score = 0
        if (name.equals("$genre Mix", ignoreCase = true)) score += 100
        if (name.startsWith(genre, ignoreCase = true)) score += 40
        if (Regex("(?i)\\b${Regex.escape(genre)}\\b").containsMatchIn(name)) score += 30
        else if (name.contains(genre, ignoreCase = true)) score += 10
        return score
    }

    /** Named playlist stations only — not “Rock Mix” / “Daily Mix” style rows. */
    fun isExplicitRadioPlaylistName(name: String): Boolean {
        if (mixLikePlaylistPattern.containsMatchIn(name) && !explicitRadioPlaylistPattern.containsMatchIn(name)) {
            return false
        }
        return explicitRadioPlaylistPattern.containsMatchIn(name)
    }

    fun historyMatchesGenre(row: StreamHistoryItem, genre: String): Boolean {
        val haystack = listOfNotNull(row.sourceLabel, row.playlist, row.album, row.artist)
        return haystack.any { it.contains(genre, ignoreCase = true) }
    }

    fun artPathForArtist(history: List<StreamHistoryItem>, artist: String): String? =
        artPathForArtistDistinct(history, artist, emptySet())

    fun artPathForGenre(history: List<StreamHistoryItem>, genre: String): String? =
        artPathForGenreDistinct(history, genre, emptySet())

    fun artPathForArtistDistinct(
        history: List<StreamHistoryItem>,
        artist: String,
        used: Set<String>,
    ): String? = history.firstOrNull { row ->
        row.filepath != null &&
            row.filepath !in used &&
            row.artist.equals(artist, ignoreCase = true)
    }?.filepath

    fun artPathForGenreDistinct(
        history: List<StreamHistoryItem>,
        genre: String,
        used: Set<String>,
    ): String? = history.firstOrNull { row ->
        row.filepath != null &&
            row.filepath !in used &&
            historyMatchesGenre(row, genre)
    }?.filepath

    fun artPathForPlaylistDistinct(
        history: List<StreamHistoryItem>,
        playlistName: String,
        used: Set<String>,
    ): String? = history.firstOrNull { row ->
        row.filepath != null &&
            row.filepath !in used &&
            row.playlist.equals(playlistName, ignoreCase = true)
    }?.filepath

    /** Next unused cover from stream history — avoids duplicate tiles when genre/artist data is sparse. */
    fun nextDistinctArtPath(history: List<StreamHistoryItem>, used: Set<String>): String? =
        history.firstOrNull { row ->
            row.filepath != null && row.filepath !in used
        }?.filepath

    fun topArtistForGenre(history: List<StreamHistoryItem>, genre: String): String? =
        history
            .filter { row -> row.artist != null && historyMatchesGenre(row, genre) }
            .groupingBy { it.artist!!.lowercase() }
            .eachCount()
            .maxByOrNull { it.value }
            ?.key
            ?.let { key -> history.firstOrNull { it.artist.equals(key, true) }?.artist }

    fun matchesKeywords(text: String, keywords: List<String>): Boolean =
        keywords.any { keyword -> text.contains(keyword, ignoreCase = true) }

    fun playlistSearchText(playlist: PlaylistSummary): String =
        listOfNotNull(playlist.name, playlist.sourceName, playlist.source).joinToString(" ")

    fun playlistMatchesTheme(playlist: PlaylistSummary, theme: HomeTheme): Boolean =
        playlistThemeScore(playlist, theme) > 0

    fun playlistMatchesTheme(name: String, theme: HomeTheme): Boolean =
        matchesKeywords(name, theme.playlistKeywords)

    fun genreMatchesTheme(name: String, theme: HomeTheme): Boolean =
        matchesKeywords(name, theme.genreKeywords)

    fun historyMatchesTheme(row: StreamHistoryItem, theme: HomeTheme): Boolean {
        val haystack = listOfNotNull(row.sourceLabel, row.playlist, row.album, row.artist)
        val keywords = theme.playlistKeywords + theme.genreKeywords
        return haystack.any { matchesKeywords(it, keywords) }
    }

    fun topArtistForTheme(history: List<StreamHistoryItem>, theme: HomeTheme): String? =
        history
            .filter { row -> row.artist != null && historyMatchesTheme(row, theme) }
            .groupingBy { it.artist!!.lowercase() }
            .eachCount()
            .maxByOrNull { it.value }
            ?.key
            ?.let { key -> history.firstOrNull { it.artist.equals(key, true) }?.artist }

    fun matchingLibraryGenre(theme: HomeTheme, libraryGenres: List<GenreItem>): String? =
        libraryGenres.firstOrNull { genreMatchesTheme(it.name, theme) }?.name

    /** Genre label from a synthetic home tile title (e.g. "Classical Era Mix" → "Classical Era"). */
    fun mixGenreLabel(title: String): String? =
        title.removeSuffix(" Mix").takeIf { it.isNotBlank() && it.length < title.length }

    /** Genre label from a genre-radio tile (e.g. "Jazz Radio" → "Jazz"). */
    fun genreRadioLabel(displayTitle: String): String? =
        displayTitle.removeSuffix(" Radio").takeIf { it.isNotBlank() && it.length < displayTitle.length }

    /** Match analytics / mix label (e.g. "Classical Era") to a library genre row for stable cover art. */
    fun matchingLibraryGenreForLabel(label: String, libraryGenres: List<GenreItem>): GenreItem? {
        val g = label.trim()
        if (g.isBlank()) return null
        libraryGenres.firstOrNull { it.name.equals(g, ignoreCase = true) }?.let { return it }
        return libraryGenres.firstOrNull {
            nameContainsGenre(it.name, g) || nameContainsGenre(g, it.name)
        }
    }

    fun playlistThemeScore(playlist: PlaylistSummary, theme: HomeTheme): Int =
        playlistThemeScore(playlistSearchText(playlist), theme)

    fun playlistThemeScore(name: String, theme: HomeTheme): Int {
        val haystack = name.lowercase()
        var score = 0
        for (keyword in theme.playlistKeywords) {
            if (haystack.contains(keyword.lowercase())) score += 10
        }
        for (keyword in theme.genreKeywords) {
            if (haystack.contains(keyword.lowercase())) score += 4
        }
        return score
    }

    fun playlistMatchesMoodSection(playlist: PlaylistSummary, theme: HomeTheme): Boolean =
        matchesKeywords(playlistSearchText(playlist), theme.playlistKeywords)

    fun playlistMatchesMoodSection(name: String, theme: HomeTheme): Boolean =
        matchesKeywords(name, theme.playlistKeywords)

    fun playlistsForMoodSection(all: List<PlaylistSummary>, theme: HomeTheme): List<PlaylistSummary> =
        all
            .filter { it.tracks > 0 && playlistMatchesMoodSection(it, theme) }
            .sortedWith(
                compareByDescending<PlaylistSummary> { playlistKeywordScore(it, theme) }
                    .thenBy { it.name.lowercase() },
            )

    private fun playlistKeywordScore(playlist: PlaylistSummary, theme: HomeTheme): Int {
        val haystack = playlistSearchText(playlist).lowercase()
        return theme.playlistKeywords.count { haystack.contains(it.lowercase()) }
    }

    fun playlistsForTheme(all: List<PlaylistSummary>, theme: HomeTheme): List<PlaylistSummary> =
        all
            .filter { it.tracks > 0 && playlistThemeScore(it, theme) > 0 }
            .sortedWith(
                compareByDescending<PlaylistSummary> { playlistThemeScore(it, theme) }
                    .thenByDescending { it.tracks },
            )

    private val decadePatterns: Map<String, List<Regex>> = mapOf(
        "60s" to listOf(
            Regex("(?i)\\b60'?s\\b"),
            Regex("(?i)'60s"),
            Regex("(?i)\\bsixties\\b"),
            Regex("(?i)\\b1960s\\b"),
            Regex("(?i)\\b196[0-9]\\b"),
        ),
        "70s" to listOf(
            Regex("(?i)\\b70'?s\\b"),
            Regex("(?i)'70s"),
            Regex("(?i)\\bseventies\\b"),
            Regex("(?i)\\b1970s\\b"),
            Regex("(?i)\\b197[0-9]\\b"),
        ),
        "80s" to listOf(
            Regex("(?i)\\b80'?s\\b"),
            Regex("(?i)'80s"),
            Regex("(?i)\\beighties\\b"),
            Regex("(?i)\\b1980s\\b"),
            Regex("(?i)\\b198[0-9]\\b"),
        ),
        "90s" to listOf(
            Regex("(?i)\\b90'?s\\b"),
            Regex("(?i)'90s"),
            Regex("(?i)\\bnineties\\b"),
            Regex("(?i)\\b1990s\\b"),
            Regex("(?i)\\b199[0-9]\\b"),
        ),
    )

    fun playlistMatchesDecade(playlist: PlaylistSummary, decadeId: String): Boolean =
        playlistMatchesDecade(playlistSearchText(playlist), decadeId)

    fun playlistMatchesDecade(name: String, decadeId: String): Boolean {
        val patterns = decadePatterns[decadeId] ?: return false
        return patterns.any { it.containsMatchIn(name) }
    }

    /** All library playlists named for this decade; multi-decade names match every listed decade. */
    fun playlistsForDecadeSection(all: List<PlaylistSummary>, decadeId: String): List<PlaylistSummary> =
        all
            .filter { it.tracks > 0 && playlistMatchesDecade(it, decadeId) }
            .sortedBy { it.name.lowercase() }

    fun isSpecialHomePlaylistName(name: String): Boolean =
        isDailyMixName(name) || isDiscoverName(name) || isGenreMixPlaylistName(name) ||
            isExplicitRadioPlaylistName(name) || isAutomationPlaylistName(name)

    /** Scheduled automation playlists — excluded from recents, mixes, and shortcut tiles. */
    fun isAutomationPlaylistName(name: String): Boolean =
        name.trim().startsWith("Automations", ignoreCase = true)

    fun browsablePlaylists(all: List<PlaylistSummary>): List<PlaylistSummary> =
        all.filter { it.tracks > 0 && !isSpecialHomePlaylistName(it.name) }

    fun shuffledBrowsablePlaylists(all: List<PlaylistSummary>, seed: Long): List<PlaylistSummary> {
        val safeSeed = if (seed == 0L) 0x4d595449L else seed
        return browsablePlaylists(all).shuffled(Random(safeSeed))
    }

    /**
     * Every playlist eligible for the home catch-all row: anything with tracks,
     * excluding scheduled automations and the server's auto-generated daily mixes
     * (those have their own row). Special genre/mix/radio names are kept so they
     * can still surface in the catch-all when their dedicated rows are full.
     */
    fun allHomePlaylists(all: List<PlaylistSummary>): List<PlaylistSummary> =
        all.filter { it.tracks > 0 && !isAutomationPlaylistName(it.name) && !isDailyMixName(it.name) }

    /** Daily-rotated full playlist set so the home teaser eventually surfaces them all. */
    fun shuffledAllPlaylists(all: List<PlaylistSummary>, seed: Long): List<PlaylistSummary> {
        val safeSeed = if (seed == 0L) 0x4d595449L else seed
        return allHomePlaylists(all).shuffled(Random(safeSeed))
    }
}

File: ios/BockMedia/Domain/HomeFeedRules.swift
import Foundation

enum HomeFeedRules {
    private static let dailyMixPattern = try! NSRegularExpression(pattern: "(?i)daily mix|daylist")
    private static let discoverPattern = try! NSRegularExpression(pattern: "(?i)discover weekly|new release|fresh find|new to you")
    private static let genreMixPattern = try! NSRegularExpression(
        pattern: "(?i)\\bmix\\b|\\bmixes\\b|\\bremix\\b|\\bremixes\\b|essentials|" +
            "\\bdecade\\b|\\bera\\b|\\bhits\\b|\\bparty\\b|\\bfocus\\b|\\bfavorites\\b"
    )
    private static let mixLikeNamePattern = try! NSRegularExpression(
        pattern: "(?i)\\bmix\\b|\\bmixes\\b|\\bremix\\b|\\bremixes\\b"
    )
    private static let explicitRadioPattern = try! NSRegularExpression(pattern: "(?i)\\bradio\\b|\\bstation\\b")
    private static let mixLikePattern = try! NSRegularExpression(pattern: "(?i)\\bmix\\b|daily|discover weekly|essentials|station")

    static func isDailyMixName(_ name: String) -> Bool {
        dailyMixPattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil
    }

    static func isDiscoverName(_ name: String) -> Bool {
        discoverPattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil
    }

    static func isGenreMixPlaylistName(_ name: String, genre: String? = nil) -> Bool {
        if isDailyMixName(name) || isDiscoverName(name) { return false }
        guard genreMixPattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil else { return false }
        guard let genre else { return true }
        return nameContainsGenre(name, genre: genre)
    }

    static func hasMixLikeName(_ name: String) -> Bool {
        mixLikeNamePattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil
    }

    static func nameContainsGenre(_ name: String, genre: String) -> Bool {
        let g = genre.trimmingCharacters(in: .whitespacesAndNewlines)
        if g.isEmpty { return false }
        if name.localizedCaseInsensitiveContains(g) { return true }
        let tokens = g.split(separator: " ").map(String.init).filter { $0.count > 1 }
        return tokens.count > 1 && tokens.all { name.localizedCaseInsensitiveContains($0) }
    }

    static func bestGenreMixPlaylist(_ all: [PlaylistSummary], genre: String) -> PlaylistSummary? {
        let g = genre.trimmingCharacters(in: .whitespacesAndNewlines)
        if g.isEmpty { return nil }
        return all.filter { $0.tracks > 0 && isGenreMixPlaylistName($0.name, genre: g) }
            .max { lhs, rhs in
                let ls = genreMixNameScore(lhs.name, genre: g) * 10_000 + lhs.tracks
                let rs = genreMixNameScore(rhs.name, genre: g) * 10_000 + rhs.tracks
                return ls < rs
            }
    }

    private static func genreMixNameScore(_ name: String, genre: String) -> Int {
        var score = 0
        if name.caseInsensitiveCompare("\(genre) Mix") == .orderedSame { score += 100 }
        if name.lowercased().hasPrefix(genre.lowercased()) { score += 40 }
        if name.range(of: "\\b\(NSRegularExpression.escapedPattern(for: genre))\\b", options: [.regularExpression, .caseInsensitive]) != nil {
            score += 30
        } else if name.localizedCaseInsensitiveContains(genre) {
            score += 10
        }
        return score
    }

    static func isExplicitRadioPlaylistName(_ name: String) -> Bool {
        let hasMixLike = mixLikePattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil
        let hasRadio = explicitRadioPattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil
        if hasMixLike && !hasRadio { return false }
        return hasRadio
    }

    static func historyMatchesGenre(_ row: StreamHistoryItem, genre: String) -> Bool {
        [row.sourceLabel, row.playlist, row.album, row.artist]
            .compactMap { $0 }
            .contains { $0.localizedCaseInsensitiveContains(genre) }
    }

    static func artPathForArtistDistinct(_ history: [StreamHistoryItem], artist: String, used: Set<String>) -> String? {
        history.first { row in
            guard let path = row.filepath, !used.contains(path) else { return false }
            return row.artist?.caseInsensitiveCompare(artist) == .orderedSame
        }?.filepath
    }

    static func artPathForGenreDistinct(_ history: [StreamHistoryItem], genre: String, used: Set<String>) -> String? {
        history.first { row in
            guard let path = row.filepath, !used.contains(path) else { return false }
            return historyMatchesGenre(row, genre: genre)
        }?.filepath
    }

    static func artPathForPlaylistDistinct(_ history: [StreamHistoryItem], playlistName: String, used: Set<String>) -> String? {
        history.first { row in
            guard let path = row.filepath, !used.contains(path) else { return false }
            return row.playlist?.caseInsensitiveCompare(playlistName) == .orderedSame
        }?.filepath
    }

    static func nextDistinctArtPath(_ history: [StreamHistoryItem], used: Set<String>) -> String? {
        history.first { row in
            guard let path = row.filepath else { return false }
            return !used.contains(path)
        }?.filepath
    }

    static func topArtistForGenre(_ history: [StreamHistoryItem], genre: String) -> String? {
        var counts: [String: Int] = [:]
        for row in history {
            guard let artist = row.artist, historyMatchesGenre(row, genre: genre) else { continue }
            counts[artist.lowercased(), default: 0] += 1
        }
        guard let top = counts.max(by: { $0.value < $1.value })?.key else { return nil }
        return history.first { $0.artist?.caseInsensitiveCompare(top) == .orderedSame }?.artist
    }

    static func matchesKeywords(_ text: String, keywords: [String]) -> Bool {
        keywords.contains { text.localizedCaseInsensitiveContains($0) }
    }

    static func playlistSearchText(_ playlist: PlaylistSummary) -> String {
        [playlist.name, playlist.sourceName, playlist.source].compactMap { $0 }.joined(separator: " ")
    }

    static func playlistMatchesTheme(_ playlist: PlaylistSummary, theme: HomeTheme) -> Bool {
        playlistThemeScore(playlist, theme: theme) > 0
    }

    static func playlistMatchesTheme(_ name: String, theme: HomeTheme) -> Bool {
        matchesKeywords(name, keywords: theme.playlistKeywords)
    }

    static func genreMatchesTheme(_ name: String, theme: HomeTheme) -> Bool {
        matchesKeywords(name, keywords: theme.genreKeywords)
    }

    static func historyMatchesTheme(_ row: StreamHistoryItem, theme: HomeTheme) -> Bool {
        let haystack = [row.sourceLabel, row.playlist, row.album, row.artist].compactMap { $0 }
        let keywords = theme.playlistKeywords + theme.genreKeywords
        return haystack.contains { matchesKeywords($0, keywords: keywords) }
    }

    static func topArtistForTheme(_ history: [StreamHistoryItem], theme: HomeTheme) -> String? {
        var counts: [String: Int] = [:]
        for row in history {
            guard let artist = row.artist, historyMatchesTheme(row, theme: theme) else { continue }
            counts[artist.lowercased(), default: 0] += 1
        }
        guard let top = counts.max(by: { $0.value < $1.value })?.key else { return nil }
        return history.first { $0.artist?.caseInsensitiveCompare(top) == .orderedSame }?.artist
    }

    static func matchingLibraryGenre(_ theme: HomeTheme, libraryGenres: [GenreItem]) -> String? {
        libraryGenres.first { genreMatchesTheme($0.name, theme: theme) }?.name
    }

    static func matchingLibraryGenreForLabel(_ label: String, libraryGenres: [GenreItem]) -> GenreItem? {
        let g = label.trimmingCharacters(in: .whitespacesAndNewlines)
        if g.isEmpty { return nil }
        if let exact = libraryGenres.first(where: { $0.name.caseInsensitiveCompare(g) == .orderedSame }) { return exact }
        return libraryGenres.first {
            nameContainsGenre($0.name, genre: g) || nameContainsGenre(g, genre: $0.name)
        }
    }

    static func playlistThemeScore(_ playlist: PlaylistSummary, theme: HomeTheme) -> Int {
        playlistThemeScore(playlistSearchText(playlist), theme: theme)
    }

    static func playlistThemeScore(_ name: String, theme: HomeTheme) -> Int {
        let haystack = name.lowercased()
        var score = 0
        for keyword in theme.playlistKeywords where haystack.contains(keyword.lowercased()) {
            score += 10
        }
        for keyword in theme.genreKeywords where haystack.contains(keyword.lowercased()) {
            score += 4
        }
        return score
    }

    static func playlistMatchesMoodSection(_ playlist: PlaylistSummary, theme: HomeTheme) -> Bool {
        matchesKeywords(playlistSearchText(playlist), keywords: theme.playlistKeywords)
    }

    static func playlistMatchesMoodSection(_ name: String, theme: HomeTheme) -> Bool {
        matchesKeywords(name, keywords: theme.playlistKeywords)
    }

    static func playlistsForMoodSection(_ all: [PlaylistSummary], theme: HomeTheme) -> [PlaylistSummary] {
        all
            .filter { $0.tracks > 0 && playlistMatchesMoodSection($0, theme: theme) }
            .sorted {
                let ls = playlistKeywordScore($0, theme: theme)
                let rs = playlistKeywordScore($1, theme: theme)
                if ls != rs { return ls > rs }
                return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
    }

    private static func playlistKeywordScore(_ playlist: PlaylistSummary, theme: HomeTheme) -> Int {
        let haystack = playlistSearchText(playlist).lowercased()
        return theme.playlistKeywords.filter { haystack.contains($0.lowercased()) }.count
    }

    static func playlistsForTheme(_ all: [PlaylistSummary], theme: HomeTheme) -> [PlaylistSummary] {
        all
            .filter { $0.tracks > 0 && playlistThemeScore($0, theme: theme) > 0 }
            .sorted {
                let ls = playlistThemeScore($0, theme: theme)
                let rs = playlistThemeScore($1, theme: theme)
                if ls != rs { return ls > rs }
                return $0.tracks > $1.tracks
            }
    }

    static func isSpecialHomePlaylistName(_ name: String) -> Bool {
        isDailyMixName(name) || isDiscoverName(name) || isGenreMixPlaylistName(name)
            || isExplicitRadioPlaylistName(name) || isAutomationPlaylistName(name)
    }

    /// Scheduled automation playlists — excluded from recents, mixes, and shortcut tiles.
    static func isAutomationPlaylistName(_ name: String) -> Bool {
        name.trimmingCharacters(in: .whitespaces).lowercased().hasPrefix("automations")
    }

    static func browsablePlaylists(_ all: [PlaylistSummary]) -> [PlaylistSummary] {
        all.filter { $0.tracks > 0 && !isSpecialHomePlaylistName($0.name) }
    }

    static func shuffledBrowsablePlaylists(_ all: [PlaylistSummary], seed: UInt64) -> [PlaylistSummary] {
        var generator = SeededRandomNumberGenerator(seed: seed)
        return browsablePlaylists(all).shuffled(using: &generator)
    }

    /// Every playlist eligible for the home catch-all row (excludes automations
    /// and the server's auto-generated daily mixes, which have their own row).
    static func allHomePlaylists(_ all: [PlaylistSummary]) -> [PlaylistSummary] {
        all.filter { $0.tracks > 0 && !isAutomationPlaylistName($0.name) && !isDailyMixName($0.name) }
    }

    /// Daily-rotated full playlist set so the home teaser eventually surfaces them all.
    static func shuffledAllPlaylists(_ all: [PlaylistSummary], seed: UInt64) -> [PlaylistSummary] {
        var generator = SeededRandomNumberGenerator(seed: seed)
        return allHomePlaylists(all).shuffled(using: &generator)
    }
}

struct SeededRandomNumberGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed == 0 ? 0x4d595449 : seed
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }
}
